R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+#' of m/z and intensity pairs than the spectrum with index `main` in `x`, also
+#' all other related information is taken from this spectrum.
+#'
+#' @author Johannes Rainer
+#'
+#' @seealso
+#'
+#' [estimateMzScattering()] for a function to estimate m/z scattering
+#' in consecutive scans.
+#'
+#' [estimateMzResolution()] for a function estimating the m/z resolution of
+#' a spectrum.
+#' 
+#' [combineSpectraMovingWindow()] for the function to combine consecutive
+#' spectra of an `MSnExp` object using a moving window approach.
+#' 
+#' @md
+#'
+#' @examples
+#'
+#' library(MSnbase)
+#' ## Create 3 example profile-mode spectra with a resolution of 0.1 and small
+#' ## random variations on these m/z values on consecutive scans.
+#' set.seed(123)
+#' mzs <- seq(1, 20, 0.1)
+#' ints1 <- abs(rnorm(length(mzs), 10))
+#' ints1[11:20] <- c(15, 30, 90, 200, 500, 300, 100, 70, 40, 20) # add peak
+#' ints2 <- abs(rnorm(length(mzs), 10))
+#' ints2[11:20] <- c(15, 30, 60, 120, 300, 200, 90, 60, 30, 23)
+#' ints3 <- abs(rnorm(length(mzs), 10))
+#' ints3[11:20] <- c(13, 20, 50, 100, 200, 100, 80, 40, 30, 20)
+#'
+#' ## Create the spectra.
+#' sp1 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.01),
+#'     intensity = ints1)
+#' sp2 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.01),
+#'     intensity = ints2)
+#' sp3 <- new("Spectrum1", mz = mzs + rnorm(length(mzs), sd = 0.009),
+#'     intensity = ints3)
+#'
+#' ## Combine the spectra
+#' sp_agg <- combineSpectra(list(sp1, sp2, sp3))
+#'
+#' ## Plot the spectra before and after combining
+#' par(mfrow = c(2, 1), mar = c(4.3, 4, 1, 1))
+#' plot(mz(sp1), intensity(sp1), xlim = range(mzs[5:25]), type = "h", col = "red")
+#' points(mz(sp2), intensity(sp2), type = "h", col = "green")
+#' points(mz(sp3), intensity(sp3), type = "h", col = "blue")
+#' plot(mz(sp_agg), intensity(sp_agg), xlim = range(mzs[5:25]), type = "h",
+#'     col = "black")
+combineSpectra <- function(x, mzFun = base::mean, intensityFun = base::mean,
+                           main = ceiling(median(1:length(x))),
+                           mzd) {
+    if (length(unique(unlist(lapply(x, function(z) z@msLevel)))) != 1)
+        stop("Can only combine spectra with the same MS level")
+    mzs <- lapply(x, function(z) z@mz)
+    mzs_lens <- base::lengths(mzs)
+    mzs <- unlist(mzs, use.names = FALSE)
+    mz_order <- base::order(mzs)
+    mzs <- mzs[mz_order]
+    mz_groups <- .group_mz_values(mzs, mzd = mzd)
+    if (length(unique(mz_groups)) < length(x[[main]]@mz))
+        stop("Got less m/z groups than m/z values in the original spectrum. ",
+             "Most likely the data is not profile-mode LCMS data.")
+    ## Want to keep only those groups with a m/z from the main spectrum.
+    keep <- unlist(lapply(
+        split(
+            rep(1:length(mzs_lens), mzs_lens)[mz_order] == main,
+            mz_groups),
+        function(z) {
+            rep(any(z), length(z))
+        }), use.names = FALSE)
```

Comment:
This could be vectorized (no need to `split` and `lapply`):
```r
    isInMain <- rep.int(seq_along(mzs_lens), mzs_lens)[mz_order] == main
    keep <- mz_groups %in% mz_groups[isInMain]
```

Reply:
Thanks! I'll amend.